### PR TITLE
[AArch64] Peephole optimization to remove redundant csel instructions

### PR DESCRIPTION
**Pull request by citymarina**
Given a sequence such as

  %8:gpr64 = COPY $xzr
  %10:gpr64 = COPY $xzr
  %11:gpr64 = CSELXr %8:gpr64, %10:gpr64, 0, implicit $nzcv

`PeepholeOptimizer::foldRedundantCopy` led to the creation of select instructions where both inputs were the same register:

  %11:gpr64 = CSELXr %8:gpr64, %8:gpr64, 0, implicit $nzcv

This change adds a later peephole optimization that replaces such selects with unconditional moves.

### CODE REVIEW

**Comment by citymarina** [comment]
The clang-format issue is with a pre-existing comment, and the suggested fix makes the code worse. Should I just mark that whole comment block (or perhaps just paragraph 7?) as `clang-format off`?

File: llvm/lib/Target/AArch64/AArch64MIPeepholeOpt.cpp [hunk 1]
```diff
@@ -283,6 +287,26 @@ bool AArch64MIPeepholeOpt::visitORR(MachineInstr &MI) {
...
+  auto ZeroReg =
+      MI.getOpcode() == AArch64::CSELXr ? AArch64::XZR : AArch64::WZR;
+  auto OrOpcode =
+      MI.getOpcode() == AArch64::CSELXr ? AArch64::ORRXrs : AArch64::ORRWrs;
+
+  BuildMI(*MI.getParent(), MI, MI.getDebugLoc(), TII->get(OrOpcode))
+      .addReg(MI.getOperand(0).getReg(), RegState::Define)
+      .addReg(ZeroReg)
+      .addReg(MI.getOperand(1).getReg())
+      .addImm(0);
```
Comment (davemgreen) [inline llvm/lib/Target/AArch64/AArch64MIPeepholeOpt.cpp:304]: I never know if we need it or not, but should this do a MRI->constrainRegClass on the Regs?

**Review by davemgreen — COMMENTED** [review]
The patch LGTM, with a question related to constraining the regclass.

> The clang-format issue is with a pre-existing comment, and the suggested fix makes the code worse. Should I just mark that whole comment block (or perhaps just paragraph 7?) as `clang-format off`?

Can you rebase over c7c5e05389292da7e5a87e3d1d3ef08021911a53, that should hopefully be enough to fix the formatting.

File: llvm/lib/Target/AArch64/AArch64MIPeepholeOpt.cpp (hunk 1 above)
Comment (citymarina) [reply llvm/lib/Target/AArch64/AArch64MIPeepholeOpt.cpp:304]: As far as I understand, the `ORR` should be ok with the same sets of registers as the `CSEL`.

**Review by davemgreen — APPROVED** [review]
OK Thanks, LGTM then.

Are you happy for me to hit submit?

**Comment by citymarina** [comment]
That would be great - thank you!

**Pull request merged**